libr/anal/sign.c

```diff
@@ -292,6 +304,34 @@ static bool addItem(RAnal *a, RSignItem *it) {
 	return retval;
 }
 
+static bool addHash(RAnal *a, const char *name, ut64 size, const ut8 *bytes) {
+	bool retval = false;
+	if (!bytes) {
+		return false;
+	}
+	
+	RSignItem *it = r_sign_item_new ();
+	if (!it) {
+		goto beach;
+	}
+	
+	it->name = r_str_new (name);
+	if (!it->name) {
+		goto beach;
+	}
+
+	char *c = r_hash_to_string (NULL, "md5", bytes, size);
```

Comment:
md5 have collisions, any reason to use this? i would probably move this to a `#define ZIGN_HASH "md5"` at the begining of the file.

Also, im not sure how much helpful a hash over the bytes of the function (whch can be non-linear) may be useful as a signature at all. You will need to compute the hash by walking all the basic blocks instead. But then again, it is probably more useful to use context-sensitive hashes rather than just md5.

Another idea would be to use the approach of common function database from bitblaze (which i have already implemented in the past, but the project is dead) and define a different prime number to each basic block, then multiply them all with bignum and use the result as a unique identifier because this way it is mathematically possible to find variations of the same function by matching just a number.

Personally i prefer this approach than just hashing the linear bytes of the function and assume this will be useful for something else than checking against the same binary

Reply:
@MattRBrooks probably can comment on this as he made at r2con. It seems that hashing is giving very good results

Reply:
@mbrooks *

Reply:
@cmatthewbrooks *

Reply:
@radare and you can probably watch his talk on the video :P

Reply:
@Maijin yes that's why I choose with MD5

Reply:
@radare I think the MD5 collision is a valid point so switched to sha256. I also had the Idea of creating an open Zignatures DB everyone can contribute, but this is probably out of scope for this Issue or can we continue here?